r1cs/lib/r1cs/Ops.cpp

```diff
@@ -0,0 +1,162 @@
+//===-- Ops.cpp - R1CS operation implementations ----------------*- C++ -*-===//
+//
+// Part of the LLZK Project, under the Apache License v2.0.
+// See LICENSE.txt for license information.
+// Copyright 2025 Veridise Inc.
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#include "r1cs/Dialect/IR/Ops.h"
+
+#include <mlir/IR/OpImplementation.h>
+
+#include <llvm/Support/Casting.h>
+
+// TableGen'd implementation files
+#define GET_OP_CLASSES
+#include "r1cs/Dialect/IR/Ops.cpp.inc"
+
+using namespace mlir;
+namespace r1cs {
+
+ParseResult CircuitDefOp::parse(OpAsmParser &parser, OperationState &result) {
+  // Parse symbol name
+  StringAttr symName;
+  if (parser.parseSymbolName(symName, "sym_name", result.attributes)) {
+    return failure();
+  }
+
+  SmallVector<OpAsmParser::Argument> args;
+  DenseMap<unsigned, Attribute> perArgAttrs;
+
+  if (succeeded(parser.parseOptionalKeyword("inputs"))) {
+    if (parser.parseLParen()) {
+      return failure();
+    }
+
+    unsigned idx = 0;
+    do {
+      OpAsmParser::Argument arg;
+      Type type;
+      if (parser.parseArgument(arg) || parser.parseColonType(type)) {
+        return failure();
+      }
+      arg.type = type;
+
+      // Try to parse an optional `{...}` attr
+      NamedAttrList attrList;
+      if (succeeded(parser.parseOptionalAttrDict(attrList))) {
+        if (auto pubAttr = attrList.get("pub")) {
+          perArgAttrs[idx] = pubAttr;
+        }
+      }
+
+      args.push_back(arg);
+      idx++;
+    } while (succeeded(parser.parseOptionalComma()));
+
+    if (parser.parseRParen()) {
+      return failure();
+    }
+  }
+
+  // Later: attach DictionaryAttr mapping "0", "1", ... to parsed attr
+  NamedAttrList attrs;
+  for (auto [i, attr] : perArgAttrs) {
+    attrs.append(std::to_string(i), attr);
+  }
+
+  if (!attrs.empty()) {
+    result.addAttribute("arg_attrs", DictionaryAttr::get(parser.getContext(), attrs));
+  }
```

Comment:
perf: check equivalent condition earlier
```suggestion
  if (!perArgAttrs.empty()) {
    NamedAttrList attrs;
    for (auto [i, attr] : perArgAttrs) {
      attrs.append(std::to_string(i), attr);
    }

    result.addAttribute("arg_attrs", DictionaryAttr::get(parser.getContext(), attrs));
  }
```